File: remote/src/main/java/com/kelsos/mbrc/adapters/NowPlayingAdapter.java
package com.kelsos.mbrc.adapters;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.TextView;
import butterknife.Bind;
import butterknife.ButterKnife;
import com.google.inject.Inject;
import com.kelsos.mbrc.R;
import com.kelsos.mbrc.dao.QueueTrackDao;
import com.kelsos.mbrc.domain.QueueTrack;
import com.kelsos.mbrc.interfaces.ITouchHelperAdapter;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import roboguice.util.Ln;

public class NowPlayingAdapter extends RecyclerView.Adapter<NowPlayingAdapter.TrackHolder>
    implements ITouchHelperAdapter {
  private List<QueueTrack> data;
  private int playingTrackIndex;
  private LayoutInflater inflater;
  private OnUserActionListener listener;

  @Inject public NowPlayingAdapter(Context context) {
    inflater = LayoutInflater.from(context);
    data = new ArrayList<>();
    setHasStableIds(true);
  }

  public void updateData(List<QueueTrack> data) {
    this.data.addAll(data);
    notifyDataSetChanged();
  }

  public void clearData() {
    this.data.clear();
    notifyDataSetChanged();
  }

  public void setPlayingTrackIndex(int index) {
    this.playingTrackIndex = index;
    notifyDataSetChanged();
  }

  @Override public TrackHolder onCreateViewHolder(ViewGroup parent, int viewType) {
    final View view = inflater.inflate(R.layout.ui_list_track_item, parent, false);
    return new TrackHolder(view);
  }

  @Override public void onBindViewHolder(TrackHolder holder, int position) {
    QueueTrack track = data.get(position);
    holder.title.setText(track.getTitle());
    holder.artist.setText(track.getArtist());

    if (position == playingTrackIndex) {
      holder.trackPlaying.setImageResource(R.drawable.ic_media_now_playing);
    } else {
      holder.trackPlaying.setImageResource(android.R.color.transparent);
    }

    holder.container.setOnClickListener(v -> {
      Ln.v("Clicked");
      if (listener == null) {
        return;
      }

      listener.onItemClicked(position, track);
    });
  }

  @Override public int getItemCount() {
    return data.size();
  }

  @Override public long getItemId(int position) {
    return data.get(position).getPosition();
  }

  public void setOnUserActionListener(OnUserActionListener listener) {
    this.listener = listener;
  }

  /**
   * This method is used to restore the tracks to their original positions in case the move
   * failed to complete.
   *
   * @param from The original position of the {@link QueueTrackDao}.
   * @param to The position the element was original moved to.
   */
  public void restorePositions(int from, int to) {
    final QueueTrack track = data.get(to);
    data.remove(track);
    data.add(from, track);
  }

  public void insert(QueueTrack track, int index) {
    data.add(index, track);
  }

  public void setPlayingTrack(QueueTrack track) {
    setPlayingTrackIndex(data.indexOf(track));
    notifyDataSetChanged();
  }

  @Override public void onItemMove(int from, int to) {
    if (from < to) {
      for (int i = from; i < to; i++) {
        Collections.swap(data, i, i + 1);
      }
    } else {
      for (int i = from; i > to; i--) {
        Collections.swap(data, i, i - 1);
      }
    }
    notifyItemMoved(from, to);

    if (listener != null) {
      listener.onItemMoved(from, to);
    }
  }

  @Override public void onItemDismiss(int position) {
    data.remove(position);
    notifyItemRemoved(position);

    if (listener != null) {
      listener.onItemRemoved(position);
    }
  }

  public interface OnUserActionListener {
    void onItemRemoved(int position);

    void onItemMoved(int from, int to);

    void onItemClicked(int position, QueueTrack track);
  }

  static class TrackHolder extends RecyclerView.ViewHolder {
    @Bind(R.id.drag_handle) View dragHandle;
    @Bind(R.id.track_title) TextView title;
    @Bind(R.id.track_artist) TextView artist;
    @Bind(R.id.track_indicator_view) ImageView trackPlaying;
    @Bind(R.id.container) FrameLayout container;

    public TrackHolder(View itemView) {
      super(itemView);
      ButterKnife.bind(this, itemView);
    }
  }
}

File: app/src/main/kotlin/com/kelsos/mbrc/ui/navigation/nowplaying/NowPlayingAdapter.kt
package com.kelsos.mbrc.ui.navigation.nowplaying

import android.app.Activity
import android.graphics.Color
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.MotionEvent.ACTION_DOWN
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import com.kelsos.mbrc.R
import com.kelsos.mbrc.content.nowplaying.NowPlayingEntity
import com.kelsos.mbrc.ui.drag.ItemTouchHelperAdapter
import com.kelsos.mbrc.ui.drag.OnStartDragListener
import com.kelsos.mbrc.ui.drag.TouchHelperViewHolder
import kotterknife.bindView
import javax.inject.Inject

class NowPlayingAdapter
@Inject
constructor(context: Activity) : RecyclerView.Adapter<NowPlayingAdapter.TrackHolder>(),
  ItemTouchHelperAdapter {

  private val dragStartListener: OnStartDragListener = context as OnStartDragListener

  private var data: List<NowPlayingEntity>? = null
  private var playingTrackIndex: Int = 0
  private var currentTrack: String = ""
  private val inflater: LayoutInflater = LayoutInflater.from(context)

  private var listener: NowPlayingListener? = null

  fun setPlayingTrack(index: Int) {
    notifyItemChanged(playingTrackIndex)
    this.playingTrackIndex = index
    notifyItemChanged(index)
  }

  fun getPlayingTrackIndex(): Int {
    return this.playingTrackIndex
  }

  fun setPlayingTrack(path: String) {
    if (data == null) {
      return
    }

    this.currentTrack = path
    data?.forEachIndexed { index, (_, _, itemPath) ->
      if (itemPath == path) {
        setPlayingTrack(index)
      }
    }
  }

  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): TrackHolder {
    val view = inflater.inflate(R.layout.ui_list_track_item, parent, false)
    val holder = TrackHolder(view)
    holder.itemView.setOnClickListener { onClick(holder) }
    holder.dragHandle.setOnTouchListener { _, motionEvent ->
      if (motionEvent.action == ACTION_DOWN) {
        dragStartListener.onStartDrag(holder)
      }
      return@setOnTouchListener false
    }
    return holder
  }

  private fun onClick(holder: TrackHolder) {
    listener?.let {
      val position = holder.adapterPosition
      setPlayingTrack(position)
      it.onPress(position)
    }
  }

  override fun onBindViewHolder(holder: TrackHolder, position: Int) {
    val nowPlaying = data?.get(holder.adapterPosition)

    nowPlaying?.let { (title, artist) ->
      holder.title.text = title
      holder.artist.text = artist
      if (position == playingTrackIndex) {
        holder.trackPlaying.setImageResource(R.drawable.ic_media_now_playing)
      } else {
        holder.trackPlaying.setImageResource(android.R.color.transparent)
      }
    }
  }

  override fun getItemCount(): Int {
    return data?.size ?: 0
  }

  override fun onItemMove(from: Int, to: Int): Boolean {
    listener?.onMove(from, to)
    notifyItemMoved(from, to)

    if (!currentTrack.isBlank()) {
      setPlayingTrack(currentTrack)
    }

    return true
  }

  override fun onItemDismiss(position: Int) {
    val nowPlaying = data?.get(position)

    nowPlaying?.let {
      //it.delete()
      notifyItemRemoved(position)
      listener?.onDismiss(position)
    }
  }

  fun update(cursor: List<NowPlayingEntity>) {
    this.data = cursor
    notifyDataSetChanged()
  }

  fun setListener(listener: NowPlayingListener) {
    this.listener = listener
  }

  interface NowPlayingListener {
    fun onPress(position: Int)
    fun onMove(from: Int, to: Int)
    fun onDismiss(position: Int)
  }

  class TrackHolder(itemView: View) : RecyclerView.ViewHolder(itemView), TouchHelperViewHolder {

    val title: TextView by bindView(R.id.track_title)
    val artist: TextView by bindView(R.id.track_artist)
    val trackPlaying: ImageView by bindView(R.id.track_indicator_view)
    val dragHandle: View by bindView(R.id.drag_handle)

    override fun onItemSelected() {
      this.itemView.setBackgroundColor(Color.DKGRAY)
    }

    override fun onItemClear() {
      this.itemView.setBackgroundColor(0)
    }
  }
}